sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3963,8 +3966,10 @@ object TimeWindowing extends Rule[LogicalPlan] {
 
         def getWindow(i: Int, dataType: DataType): Expression = {
           val timestamp = PreciseTimestampConversion(window.timeColumn, dataType, LongType)
-          val lastStart = timestamp - (timestamp - window.startTime
-            + window.slideDuration) % window.slideDuration
+          val remainder = (timestamp - window.startTime) % window.slideDuration
```

Comment:
When does processing time become negative -- is this a more fundamental problem to fix?

Also, does not seem to be a reason to use CaseWhen here? just do this in Scala code

Reply:
Thank you for reviewing.
My understanding is that whether the event time is negative or not is determined by the upstream data source, not the Spark platform itself. Even if the event time is negative, the platform should give the correct value.

Does `in scala code` refer to the following code:
```
val lastStart = if (remainder < 0) {
    timestamp - remainder - window.slideDuration
 } else {
    timestamp - remainder
}
```
In that case,The remainder is a expressions. Predicate, the remainder < 0 can't get a Boolean value.